soci/soci_index.go

```diff
@@ -333,24 +333,46 @@ func (b *IndexBuilder) Build(ctx context.Context, img images.Image) (*IndexWithM
 		return nil, err
 	}
 
-	sociLayersDesc := make([]*ocispec.Descriptor, len(manifest.Layers))
-	eg, ctx := errgroup.WithContext(ctx)
-	for i, l := range manifest.Layers {
-		i, l := i, l
-		eg.Go(func() error {
+	sociLayersDesc := make([]*ocispec.Descriptor, 0, len(manifest.Layers))
+	var mu sync.Mutex
+	var wg sync.WaitGroup
+	errChan := make(chan error, len(manifest.Layers))
+
+	for _, l := range manifest.Layers {
+		wg.Add(1)
+		go func(l ocispec.Descriptor) {
+			defer wg.Done()
 			desc, err := b.buildSociLayer(ctx, l)
+			if err == errUnsupportedLayerFormat {
+				return
+			}
 			if err != nil {
-				return fmt.Errorf("could not build zTOC for layer %s: %w", l.Digest.String(), err)
+				errChan <- err
+				return
 			}
-			sociLayersDesc[i] = desc
-			return nil
-		})
+			mu.Lock()
+			sociLayersDesc = append(sociLayersDesc, desc)
+			mu.Unlock()
+		}(l)
 	}
-	if err := eg.Wait(); err != nil {
-		return nil, err
+
+	wg.Wait()
+
+errLoop:
+	for {
+		select {
+		case err := <-errChan:
+			fmt.Print(err)
+		default:
+			break errLoop
+		}
```

Comment:
How does this section of code work? E.g., when will the `for loop` and `break errLoop` ends and the code proceeds to, e.g., line 371

Reply:
will it break and proceed, or break and go to errLoop? (what the `errLoop` label does?)

Reply:
It will break and proceed. Without the `errLoop:` label, the break statement would break out of `select`, but continue to loop in the `for`.

See https://go.dev/doc/effective_go#switch